### ISSUE TITLE
Crash caused by fences with nvidia GPU

### ISSUE THREAD

**Issue by xytovl** [WiVRn/WiVRn#180]
According to user reports, nvidia GPU may crash the wivrn-server process.

Current workaround is to install [Monado vulkan layers](https://gitlab.freedesktop.org/monado/utilities/vulkan-layers) to forcibly enable timeline semaphores.
* Arch linux: [AUR package](https://aur.archlinux.org/packages/monado-vulkan-layers-git)

**Comment by damaestro** [comment]
* Fedora: [monado-vulkan-layers](https://packages.fedoraproject.org/pkgs/monado-vulkan-layers/monado-vulkan-layers/)

**Comment by GamerKing122** [comment]
what about debian base?

**Comment by galister** [comment]
This is supposedly fixed with Nvidia driver version 565.77. Those who were affected, please confirm

**Comment by damaestro** [comment]
The following is working without the layers installed:

```
$ vulkaninfo --summary
==========
VULKANINFO
==========

Vulkan Instance Version: 1.3.296


Instance Extensions: count = 24
-------------------------------
VK_EXT_acquire_drm_display             : extension revision 1
VK_EXT_acquire_xlib_display            : extension revision 1
VK_EXT_debug_report                    : extension revision 10
VK_EXT_debug_utils                     : extension revision 2
VK_EXT_direct_mode_display             : extension revision 1
VK_EXT_display_surface_counter         : extension revision 1
VK_EXT_headless_surface                : extension revision 1
VK_EXT_surface_maintenance1            : extension revision 1
VK_EXT_swapchain_colorspace            : extension revision 4
VK_KHR_device_group_creation           : extension revision 1
VK_KHR_display                         : extension revision 23
VK_KHR_external_fence_capabilities     : extension revision 1
VK_KHR_external_memory_capabilities    : extension revision 1
VK_KHR_external_semaphore_capabilities : extension revision 1
VK_KHR_get_display_properties2         : extension revision 1
VK_KHR_get_physical_device_properties2 : extension revision 2
VK_KHR_get_surface_capabilities2       : extension revision 1
VK_KHR_portability_enumeration         : extension revision 1
VK_KHR_surface                         : extension revision 25
VK_KHR_surface_protected_capabilities  : extension revision 1
VK_KHR_wayland_surface                 : extension revision 6
VK_KHR_xcb_surface                     : extension revision 6
VK_KHR_xlib_surface                    : extension revision 6
VK_LUNARG_direct_driver_loading        : extension revision 1

Instance Layers: count = 7
--------------------------
VK_LAYER_KHRONOS_validation       Khronos Validation Layer     1.3.296  version 1
VK_LAYER_MESA_device_select       Linux device selection layer 1.3.211  version 1
VK_LAYER_NV_optimus               NVIDIA Optimus layer         1.3.289  version 1
VK_LAYER_VALVE_steam_fossilize_32 Steam Pipeline Caching Layer 1.3.207  version 1
VK_LAYER_VALVE_steam_fossilize_64 Steam Pipeline Caching Layer 1.3.207  version 1
VK_LAYER_VALVE_steam_overlay_32   Steam Overlay Layer          1.3.207  version 1
VK_LAYER_VALVE_steam_overlay_64   Steam Overlay Layer          1.3.207  version 1

Devices:
========
GPU0:
	apiVersion         = 1.3.289
	driverVersion      = 565.77.0.0
	vendorID           = 0x10de
	deviceID           = 0x2208
	deviceType         = PHYSICAL_DEVICE_TYPE_DISCRETE_GPU
	deviceName         = NVIDIA GeForce RTX 3080 Ti
	driverID           = DRIVER_ID_NVIDIA_PROPRIETARY
	driverName         = NVIDIA
	driverInfo         = 565.77
	conformanceVersion = 1.3.8.2
	deviceUUID         = 92fede6d-c821-3d0d-cf70-b721c965ed35
	driverUUID         = 5d948742-de2b-5e32-9692-c2a5621aed9a
GPU1:
	apiVersion         = 1.3.289
	driverVersion      = 0.0.1
	vendorID           = 0x10005
	deviceID           = 0x0000
	deviceType         = PHYSICAL_DEVICE_TYPE_CPU
	deviceName         = llvmpipe (LLVM 19.1.0, 256 bits)
	driverID           = DRIVER_ID_MESA_LLVMPIPE
	driverName         = llvmpipe
	driverInfo         = Mesa 24.2.8 (LLVM 19.1.0)
	conformanceVersion = 1.3.1.1
	deviceUUID         = 6d657361-3234-2e32-2e38-000000000000
	driverUUID         = 6c6c766d-7069-7065-5555-494400000000
```

Packages:
```
wivrn-0.22-1.20241206gitv0.22.fc41.x86_64
opencomposite-0.0.1-1.20241111git34311da.fc41.x86_64
xorg-x11-drv-nvidia-565.77-3.fc41.x86_64
```

Tested:

* hl:a via steam
* unreal engine 5.4.4 native
* xrgears

**Comment by galister** [comment]
Amazing, thanks. Going to close this. 

TLDR;
The solution is to upgrade nvidia driver to version 565.77 or later.

**Comment by 123ldk** [comment]
I have updated to nvidia driver 565.77 (and attempted to install the monando vulkan layer, I'm not sure if I did it right honestly, I built it with cmake) but neither has resolved the issue.

I'm using the flatpak version of the app, here is my system info:

OS: Pop!_OS 22.04 LTS x86_64 
Kernel: 6.9.3-76060903-generic 
CPU: 11th Gen Intel i7-11800H (16) @ 4.600GHz 
GPU: NVIDIA GeForce RTX 3070 Mobile / Max-Q 
GPU: Intel TigerLake-H GT1 [UHD Graphics] 

Here is my vulkaninfo (I have a few errors I don't quite understand)
`$ vulkaninfo --summary
WARNING: [Loader Message] Code 0 : loader_scan_for_direct_drivers: The VK_LUNARG_direct_driver_loading extension was enabled but the pNext chain of VkInstanceCreateInfo did not contain the VkDirectDriverLoadingListLUNARG structure.
WARNING: [Loader Message] Code 0 : Layer VK_LAYER_MND_enable_timeline_semaphore uses API version 1.2 which is older than the application specified API version of 1.3. May cause issues.
ERROR: [Loader Message] Code 0 : libVkLayer_MND_enable_timeline_semaphore.so: cannot open shared object file: No such file or directory
WARNING: [Loader Message] Code 0 : terminator_CreateInstance: Received return code -3 from call to vkCreateInstance in ICD /usr/lib/x86_64-linux-gnu/libvulkan_virtio.so. Skipping this driver.
==========
VULKANINFO
==========

Vulkan Instance Version: 1.3.280


Instance Extensions: count = 23
-------------------------------
VK_EXT_acquire_drm_display             : extension revision 1
VK_EXT_acquire_xlib_display            : extension revision 1
VK_EXT_debug_report                    : extension revision 10
VK_EXT_debug_utils                     : extension revision 2
VK_EXT_direct_mode_display             : extension revision 1
VK_EXT_display_surface_counter         : extension revision 1
VK_EXT_surface_maintenance1            : extension revision 1
VK_EXT_swapchain_colorspace            : extension revision 4
VK_KHR_device_group_creation           : extension revision 1
VK_KHR_display                         : extension revision 23
VK_KHR_external_fence_capabilities     : extension revision 1
VK_KHR_external_memory_capabilities    : extension revision 1
VK_KHR_external_semaphore_capabilities : extension revision 1
VK_KHR_get_display_properties2         : extension revision 1
VK_KHR_get_physical_device_properties2 : extension revision 2
VK_KHR_get_surface_capabilities2       : extension revision 1
VK_KHR_portability_enumeration         : extension revision 1
VK_KHR_surface                         : extension revision 25
VK_KHR_surface_protected_capabilities  : extension revision 1
VK_KHR_wayland_surface                 : extension revision 6
VK_KHR_xcb_surface                     : extension revision 6
VK_KHR_xlib_surface                    : extension revision 6
VK_LUNARG_direct_driver_loading        : extension revision 1

Instance Layers: count = 7
--------------------------
VK_LAYER_MESA_device_select            Linux device selection layer           1.3.211  version 1
VK_LAYER_MND_enable_timeline_semaphore Monado enable timeline semaphore layer 1.2.199  version 1
VK_LAYER_NV_optimus                    NVIDIA Optimus layer                   1.3.289  version 1
VK_LAYER_VALVE_steam_fossilize_32      Steam Pipeline Caching Layer           1.3.207  version 1
VK_LAYER_VALVE_steam_fossilize_64      Steam Pipeline Caching Layer           1.3.207  version 1
VK_LAYER_VALVE_steam_overlay_32        Steam Overlay Layer                    1.3.207  version 1
VK_LAYER_VALVE_steam_overlay_64        Steam Overlay Layer                    1.3.207  version 1

Devices:
========
GPU0:
	apiVersion         = 4206866 (1.3.274)
	driverVersion      = 100663299 (0x6000003)
	vendorID           = 0x8086
	deviceID           = 0x9a60
	deviceType         = PHYSICAL_DEVICE_TYPE_INTEGRATED_GPU
	deviceName         = Intel(R) UHD Graphics (TGL GT1)
	driverID           = DRIVER_ID_INTEL_OPEN_SOURCE_MESA
	driverName         = Intel open-source Mesa driver
	driverInfo         = Mesa 24.0.3-1pop1~1711635559~22.04~7a9f319
	conformanceVersion = 1.3.6.0
	deviceUUID         = 8680609a-0100-0000-0002-000000000000
	driverUUID         = 8d28b9ea-5748-10f9-8acd-248b54de1cf7
GPU1:
	apiVersion         = 4206881 (1.3.289)
	driverVersion      = 2371043328 (0x8d534000)
	vendorID           = 0x10de
	deviceID           = 0x249d
	deviceType         = PHYSICAL_DEVICE_TYPE_DISCRETE_GPU
	deviceName         = NVIDIA GeForce RTX 3070 Laptop GPU
	driverID           = DRIVER_ID_NVIDIA_PROPRIETARY
	driverName         = NVIDIA
	driverInfo         = 565.77
	conformanceVersion = 1.3.8.2
	deviceUUID         = c70d6799-7561-d811-a32d-3a68fb6090ac
	driverUUID         = 5d948742-de2b-5e32-9692-c2a5621aed9a
GPU2:
	apiVersion         = 4206866 (1.3.274)
	driverVersion      = 1 (0x0001)
	vendorID           = 0x10005
	deviceID           = 0x0000
	deviceType         = PHYSICAL_DEVICE_TYPE_CPU
	deviceName         = llvmpipe (LLVM 15.0.7, 256 bits)
	driverID           = DRIVER_ID_MESA_LLVMPIPE
	driverName         = llvmpipe
	driverInfo         = Mesa 24.0.3-1pop1~1711635559~22.04~7a9f319 (LLVM 15.0.7)
	conformanceVersion = 1.3.1.1
	deviceUUID         = 6d657361-3234-2e30-2e33-2d31706f7000
	driverUUID         = 6c6c766d-7069-7065-5555-494400000000
`

**Comment by 123ldk** [comment]
I have updated to nvidia driver 565.77 (and attempted to install the monando vulkan layer, I'm not sure if I did it right honestly, I built it with cmake) but neither has resolved the issue.

I'm using the flatpak version of the app, here is my system info:

OS: Pop!_OS 22.04 LTS x86_64 
Kernel: 6.9.3-76060903-generic 
CPU: 11th Gen Intel i7-11800H (16) @ 4.600GHz 
GPU: NVIDIA GeForce RTX 3070 Mobile / Max-Q 
GPU: Intel TigerLake-H GT1 [UHD Graphics] 

Here is my vulkaninfo (I have a few errors I don't quite understand)

```
$ vulkaninfo --summary
WARNING: [Loader Message] Code 0 : loader_scan_for_direct_drivers: The VK_LUNARG_direct_driver_loading extension was enabled but the pNext chain of VkInstanceCreateInfo did not contain the VkDirectDriverLoadingListLUNARG structure.
WARNING: [Loader Message] Code 0 : Layer VK_LAYER_MND_enable_timeline_semaphore uses API version 1.2 which is older than the application specified API version of 1.3. May cause issues.
ERROR: [Loader Message] Code 0 : libVkLayer_MND_enable_timeline_semaphore.so: cannot open shared object file: No such file or directory
WARNING: [Loader Message] Code 0 : terminator_CreateInstance: Received return code -3 from call to vkCreateInstance in ICD /usr/lib/x86_64-linux-gnu/libvulkan_virtio.so. Skipping this driver.
==========
VULKANINFO
==========

Vulkan Instance Version: 1.3.280


Instance Extensions: count = 23
-------------------------------
VK_EXT_acquire_drm_display             : extension revision 1
VK_EXT_acquire_xlib_display            : extension revision 1
VK_EXT_debug_report                    : extension revision 10
VK_EXT_debug_utils                     : extension revision 2
VK_EXT_direct_mode_display             : extension revision 1
VK_EXT_display_surface_counter         : extension revision 1
VK_EXT_surface_maintenance1            : extension revision 1
VK_EXT_swapchain_colorspace            : extension revision 4
VK_KHR_device_group_creation           : extension revision 1
VK_KHR_display                         : extension revision 23
VK_KHR_external_fence_capabilities     : extension revision 1
VK_KHR_external_memory_capabilities    : extension revision 1
VK_KHR_external_semaphore_capabilities : extension revision 1
VK_KHR_get_display_properties2         : extension revision 1
VK_KHR_get_physical_device_properties2 : extension revision 2
VK_KHR_get_surface_capabilities2       : extension revision 1
VK_KHR_portability_enumeration         : extension revision 1
VK_KHR_surface                         : extension revision 25
VK_KHR_surface_protected_capabilities  : extension revision 1
VK_KHR_wayland_surface                 : extension revision 6
VK_KHR_xcb_surface                     : extension revision 6
VK_KHR_xlib_surface                    : extension revision 6
VK_LUNARG_direct_driver_loading        : extension revision 1

Instance Layers: count = 7
--------------------------
VK_LAYER_MESA_device_select            Linux device selection layer           1.3.211  version 1
VK_LAYER_MND_enable_timeline_semaphore Monado enable timeline semaphore layer 1.2.199  version 1
VK_LAYER_NV_optimus                    NVIDIA Optimus layer                   1.3.289  version 1
VK_LAYER_VALVE_steam_fossilize_32      Steam Pipeline Caching Layer           1.3.207  version 1
VK_LAYER_VALVE_steam_fossilize_64      Steam Pipeline Caching Layer           1.3.207  version 1
VK_LAYER_VALVE_steam_overlay_32        Steam Overlay Layer                    1.3.207  version 1
VK_LAYER_VALVE_steam_overlay_64        Steam Overlay Layer                    1.3.207  version 1

Devices:
========
GPU0:
	apiVersion         = 4206866 (1.3.274)
	driverVersion      = 100663299 (0x6000003)
	vendorID           = 0x8086
	deviceID           = 0x9a60
	deviceType         = PHYSICAL_DEVICE_TYPE_INTEGRATED_GPU
	deviceName         = Intel(R) UHD Graphics (TGL GT1)
	driverID           = DRIVER_ID_INTEL_OPEN_SOURCE_MESA
	driverName         = Intel open-source Mesa driver
	driverInfo         = Mesa 24.0.3-1pop1~1711635559~22.04~7a9f319
	conformanceVersion = 1.3.6.0
	deviceUUID         = 8680609a-0100-0000-0002-000000000000
	driverUUID         = 8d28b9ea-5748-10f9-8acd-248b54de1cf7
GPU1:
	apiVersion         = 4206881 (1.3.289)
	driverVersion      = 2371043328 (0x8d534000)
	vendorID           = 0x10de
	deviceID           = 0x249d
	deviceType         = PHYSICAL_DEVICE_TYPE_DISCRETE_GPU
	deviceName         = NVIDIA GeForce RTX 3070 Laptop GPU
	driverID           = DRIVER_ID_NVIDIA_PROPRIETARY
	driverName         = NVIDIA
	driverInfo         = 565.77
	conformanceVersion = 1.3.8.2
	deviceUUID         = c70d6799-7561-d811-a32d-3a68fb6090ac
	driverUUID         = 5d948742-de2b-5e32-9692-c2a5621aed9a
GPU2:
	apiVersion         = 4206866 (1.3.274)
	driverVersion      = 1 (0x0001)
	vendorID           = 0x10005
	deviceID           = 0x0000
	deviceType         = PHYSICAL_DEVICE_TYPE_CPU
	deviceName         = llvmpipe (LLVM 15.0.7, 256 bits)
	driverID           = DRIVER_ID_MESA_LLVMPIPE
	driverName         = llvmpipe
	driverInfo         = Mesa 24.0.3-1pop1~1711635559~22.04~7a9f319 (LLVM 15.0.7)
	conformanceVersion = 1.3.1.1
	deviceUUID         = 6d657361-3234-2e30-2e33-2d31706f7000
	driverUUID         = 6c6c766d-7069-7065-5555-494400000000


```

**Comment by galister** [comment]
what is the crash message you are getting when trying to launch a game?

**Comment by amoshydra** [comment]
I am able to run WiVRn 0.22 (test with both AppImage and with Envision) 
I do not have Monado vulkan layers installed.

```
==========
VULKANINFO
==========

Vulkan Instance Version: 1.4.303

...


Devices:
========
GPU0:
	apiVersion         = 1.3.289
	driverVersion      = 565.77.0.0
	vendorID           = 0x10de
	deviceID           = 0x2684
	deviceType         = PHYSICAL_DEVICE_TYPE_DISCRETE_GPU
	deviceName         = NVIDIA GeForce RTX 4090
	driverID           = DRIVER_ID_NVIDIA_PROPRIETARY
	driverName         = NVIDIA
	driverInfo         = 565.77
```

**Comment by galister** [comment]
thanks, going to close this one

**Comment by shermansiu** [comment]
FYI, I cloned https://gitlab.freedesktop.org/monado/utilities/vulkan-layers and was able to run `sudo make install`, but it still says that the Monado NVIDIA layer is not installed.

I'm on Ubuntu 24.
